[24.0 backport] ci: extract buildkit version correctly with replace-d modules

thaJeztah opened the pull request:
- backport of https://github.com/moby/moby/pull/45994
- backport of https://github.com/moby/moby/pull/46012

---

**- What I did**

Extract version from replace-d buildkit module, to ensure that the right version of buildkit is fetched during tests.

This is useful for working on testing/temporary forks, as in https://github.com/moby/moby/pull/45936.

**- How I did it**

**- How to verify it**

See https://github.com/moby/moby/pull/45936.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

![image](https://github.com/moby/moby/assets/7352848/4c2852dc-ce40-4b8b-a143-c96025ccf9a5)

cc @crazy-max

jedevc committed:
ci: extract buildkit version correctly with replace-d modules

Signed-off-by: Justin Chadwell <me@jedevc.com>
(cherry picked from commit f8c0d92a22bad004cb9cbb4db704495527521c42)
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

neersighted committed:
ci(buildkit): remove misleading code from buildkit-ref

Post-f8c0d92a22bad004cb9cbb4db704495527521c42, BUILDKIT_REPO doesn't
really do what it claims to. Instead, don't allow overloading since the
import path for BuildKit is always the same, and make clear the
provenance of values when generating the final variable definitions.

We also better document the script, and follow some best practices for
both POSIX sh and Bash.

Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>
(cherry picked from commit 4ecc01f3ad56e608d32d141e8fc67565c195ec4e)
Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

neersighted reviewed (changes requested):
Blocking on https://github.com/moby/moby/pull/46012 as I think these should go in together.

neersighted committed:
hack/with-go-mod.sh: warn on stderr

Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>
(cherry picked from commit 48ff8a95cc53ba1315705da360317e7fdd4970a7)
Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

neersighted committed:
ci(buildkit): remove early-return from buildkit-ref

This doesn't really make sense now that this script returns a
$GITHUB_ENV snippet.

Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>
(cherry picked from commit 7310a7cd0ce75b104295905aad341bb50bc40a71)
Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

neersighted reviewed (approved):
Updated with my follow-up which I think should land at the same time.

LGTM but is a partial self-review.

thaJeztah reviewed (commented):
LGTM if green
